packages/agreement-process/src/services/agreementService.ts

```diff
@@ -277,3 +346,160 @@ export async function updateAgreementLogic({
+  assertTenantExist(authData.organizationId, tenant);
+  assertAgreementExist(agreementId, agreementToBeUpgraded);
+  assertRequesterIsConsumer(
+    agreementToBeUpgraded.data.consumerId,
+    authData.organizationId
+  );
+
+  const upgradableStates: AgreementState[] = [
+    agreementState.active,
+    agreementState.suspended,
+  ];
+
+  assertExpectedState(
+    agreementId,
+    agreementToBeUpgraded.data.state,
+    upgradableStates
+  );
+
+  const eservice = await eserviceQuery.getEServiceById(
+    agreementToBeUpgraded.data.eserviceId
+  );
+  assertEServiceExist(agreementToBeUpgraded.data.eserviceId, eservice);
+
+  const newDescriptor = eservice.data.descriptors.find(
+    (d) => d.state === descriptorState.published
+  );
+  if (newDescriptor === undefined) {
+    throw publishedDescriptorNotFound(agreementToBeUpgraded.data.eserviceId);
+  }
+  const latestDescriptorVersion = z
+    .preprocess((x) => Number(x), z.number())
+    .safeParse(newDescriptor.version);
+  if (!latestDescriptorVersion.success) {
+    throw unexpectedVersionFormat(eservice.data.id, newDescriptor.id);
+  }
+
+  const currentDescriptor = eservice.data.descriptors.find(
+    (d) => d.id === agreementToBeUpgraded.data.descriptorId
+  );
+  if (currentDescriptor === undefined) {
+    throw descriptorNotFound(
+      eservice.data.id,
+      agreementToBeUpgraded.data.descriptorId
+    );
+  }
+
+  const currentVersion = z
+    .preprocess((x) => Number(x), z.number())
+    .safeParse(currentDescriptor.version);
+  if (!currentVersion.success) {
+    throw unexpectedVersionFormat(eservice.data.id, currentDescriptor.id);
+  }
+
+  if (latestDescriptorVersion.data <= currentVersion.data) {
+    throw noNewerDescriptor(eservice.data.id, currentDescriptor.id);
+  }
+
+  validateCertifiedAttributes(newDescriptor, tenant.data);
+
+  const verifiedValid = verifiedAttributesSatisfied(
+    agreementToBeUpgraded.data.producerId,
+    newDescriptor,
+    tenant.data
+  );
+
+  const declaredValid = declaredAttributesSatisfied(newDescriptor, tenant.data);
+
+  if (verifiedValid && declaredValid) {
+    // upgradeAgreement
+    const stamp: AgreementStamp = {
+      who: authData.organizationId,
+      when: new Date(),
+    };
+    const deactivated: Agreement = {
+      ...agreementToBeUpgraded.data,
+      state: agreementState.archived,
+      stamps: {
+        ...agreementToBeUpgraded.data.stamps,
+        archiving: stamp,
+      },
+    };
+    const upgraded: Agreement = {
+      ...agreementToBeUpgraded.data,
+      id: uuidv4(),
+      descriptorId: newDescriptor.id,
+      createdAt: new Date(),
+      updatedAt: undefined,
+      rejectionReason: undefined,
+      stamps: {
+        ...agreementToBeUpgraded.data.stamps,
+        upgrade: stamp,
+      },
+    };
+
+    return [
+      toCreateEventAgreementUpdated(
+        deactivated,
+        agreementToBeUpgraded.metadata.version
+      ),
+      toCreateEventAgreementAdded(upgraded),
+    ];
```

Comment:
extract to function `upgradeAgreement` ? 
maybe common between other agreement service logics...

Reply:
this is the same of `createNewDraftAgreement`